execution/executor/src/lib.rs

```diff
@@ -40,29 +42,227 @@ use std::{
+        if let Some(synced_tree_state) = startup_info.synced_tree_state {
+            cache.update_synced_trees(ExecutedTrees::from(synced_tree_state));
+        }
+        cache
+    }
+
+    fn committed_block_id(&self) -> HashValue {
+        self.committed_block_id
+    }
+
+    fn committed_trees(&self) -> &ExecutedTrees {
+        &self.committed_trees
+    }
+
+    fn synced_trees(&self) -> &ExecutedTrees {
+        &self.synced_trees
+    }
+
+    fn update_block_tree_root(
+        &mut self,
+        committed_trees: ExecutedTrees,
+        committed_ledger_info: &LedgerInfo,
+    ) {
+        let new_root_block_id = if committed_ledger_info.next_validator_set().is_some() {
+            // Update the root block id with reconfig virtual block id, to be consistent
+            // with the logic of Consensus.
+            let id = Block::<()>::make_genesis_block_from_ledger_info(committed_ledger_info).id();
+            debug!(
+                "Updated with a new root block {} as a virtual block of reconfiguration block {}",
+                id,
+                committed_ledger_info.consensus_block_id()
+            );
+            id
+        } else {
+            let id = committed_ledger_info.consensus_block_id();
+            debug!("updated with a new root block {}", id);
+            id
+        };
+        self.committed_block_id = new_root_block_id;
+        self.committed_trees = committed_trees.clone();
+        self.synced_trees = committed_trees;
+    }
+
+    fn update_synced_trees(&mut self, new_trees: ExecutedTrees) {
+        self.synced_trees = new_trees;
+    }
+
+    fn reset(&mut self) {
+        self.heads = vec![];
+        *self.block_map.lock().unwrap() = HashMap::new();
+    }
+
+    fn add_block(
+        &mut self,
+        parent_block_id: HashValue,
+        block: (
+            HashValue, /* block_id */
+            Vec<Transaction>,
+            ProcessedVMOutput,
+        ),
+    ) -> Result<()> {
+        // 0. Check existence first
+        let (block_id, txns, output) = block;
+        if self.block_map.lock().unwrap().contains_key(&block_id) {
+            return Ok(());
+        }
+
+        let block = Arc::new(Mutex::new(SpeculationBlock::new(
+            block_id,
+            txns,
+            output,
+            Arc::clone(&self.block_map),
+        )));
+        // 1. Add to the map
+        self.block_map
+            .lock()
+            .unwrap()
+            .insert(block_id, Arc::downgrade(&block));
+        // 2. Add to the tree
+        if parent_block_id == self.committed_block_id() {
+            self.heads.push(block);
+            return Ok(());
+        } else {
+            self.get_block(&parent_block_id)?
+                .lock()
+                .unwrap()
+                .add_child(block);
+        }
+        Ok(())
+    }
+
+    fn prune(&mut self, committed_ledger_info: &LedgerInfo) -> Result<()> {
+        let arc_latest_committed_block =
```

Comment:
maybe instead we sort these by version and we eliminate all versions earlier than the newly committed version? alternatively it isn't really clear to me why we're cloning if we're pruning.

Reply:
I didn't get your question. I use Weak pointer in the tree so after update the `heads`, reference counting of all to-be-pruned branches will be cleaned automatically by reference counting.